### ISSUE TITLE
kubelet refresh times for configmaps is long and random

### ISSUE THREAD

**Issue by thockin** [kubernetes/kubernetes#30189]
The refreshing of config maps is coupled to kubelet's internal timers.  This means that it can be difficult to rely on the configmap (and secret) self-refresh behavior.

We should keep a list of secrets and configmaps and either watch them in kubelet or else poll them on a faster and more well-documented schedule.  Some folks don't want to depend on it as-is.

**Comment by mwitkow** [comment]
@thockin we kinda got bitten by this. Can you provide a pointer to which timers these are so we can infere the internal behaviour and adjust our stuff to it? :) thanks

**Comment by thockin** [comment]
@dchen1107 @yujuhong @timstclair I don't know where these numbers are buried any more :)

**Comment by yujuhong** [comment]
kubelet stopped syncing all pods frequently (10s) and unconditionally since v1.2, because of the  serious overhead which became increasingly evident as the pod density on the node grew. Pod syncing is now triggered mainly by:
 1. pod spec changes
 2. container state changes
 3. A long, fallback resync interval (`--sync-frequency`) defaulted to 1 minute

You could increase the resync interval, but be warned that it has the adverse effect of performance overhead if there are many pods on your node. As the issue suggests, kubelet should detect the config map changes another way.

**Comment by champtar** [comment]
@wojtek-t is talking about watch based configmap / secrets update here https://github.com/kubernetes/kubernetes/pull/46470#issuecomment-305094034
Any news about that?
Is there any way to force a config map sync today (k8s 1.7)?

**Comment by wojtek-t** [comment]
> @wojtek-t is talking about watch based configmap / secrets update here #46470 (comment)
Any news about that?

This is blocked on bulk-watch feature that is under implementation. potentially Alpha in 1.9, but without promises.

> Is there any way to force a config map sync today (k8s 1.7)?

I don't think you can force it. But it should automatically sync periodically (by default every 1m). If it's not, then there is some bug.

**Comment by champtar** [comment]
Thanks for the update @wojtek-t 
The periodic sync does work, I was just looking for instant sync without doing the watch myself in the container (using kubernetes API)
What tickets should I watch for bulk-watch feature?

**Comment by discordianfish** [comment]
Am I suppose to open a new issue now or what?
I'm so frustrated with this bot by now.. It's a valid issue and there is no fix for it. If it's a 'works as designed', that's fine. But just getting idle-closed is rude.

**Comment by dims** [comment]
should be `open` and `frozen` now @discordianfish

**Comment by discordianfish** [comment]
Thanks @dims, you are always to quick to alleviate my frustration! <3

**Comment by yujuhong** [comment]
@newtonkishore nothing stops you from configuring the --sync-frequency yourself today.

That said, @wojtek-t's configmap/secret watch PR has been merged. After it soaked for a while, we can start improving this part of the logic in kubelet. For example, we can trigger pods workers to sync the pod whenever changes are detected.

**Comment by FengyunPan2** [comment]
@yujuhong
Can you tell me the PR which watches configmap/secret? I want to find the PR.

**Comment by yujuhong** [comment]
@FengyunPan2 #64752

**Comment by warmchang** [comment]
Since the https://github.com/kubernetes/kubernetes/pull/64752 has been merged, can we close this?

**Comment by Pingan2017** [comment]
> After it soaked for a while, we can start improving this part of the logic in kubelet. For example, we can trigger pods workers to sync the pod whenever changes are detected.

Have any plan for this? @yujuhong

**Comment by jgonera** [comment]
Any news on this front? Is triggering pod workers to sync whenever changes are detected a bigger patch?

**Comment by excavador** [comment]
What the status of the issue? This is so annoying behaviour

**Comment by sauragrawal** [comment]
Any update on this issue?

**Comment by b675987273** [comment]
in pkg/kublet/kubelet.go 
[// sync interval is defaulted to 10s.](https://github.com/kubernetes/kubernetes/blob/f7091992c04e2d4f16f6b462d4a711e5b9eb0f22/pkg/kubelet/kubelet.go#L1824)
how can i change the  sync interval or find it?

**Comment by ThoTischner** [comment]
Hey have a look at https://github.com/kubernetes/website/pull/18082/files

**Comment by b675987273** [comment]
Oh, thanks. I had found it from the source code.

**Comment by ThoTischner** [comment]
> Oh, thanks. I had found it from the source code.
> BUT, why is 10s coded in pkg/kublet/kubelet.go but 1min in your files

You mean in the docs ?
The time values in the docs were already present.
The PR adds a hint for the force refresh by changing the pods annotation.

**Comment by b675987273** [comment]
got it. i have no more questions

**Comment by tangzhankun** [comment]
@ThoTischner 
From the document, it's still not very clear.
`You can trigger an immediate refresh by updating one of the pod's annotations`

For instance, per my testing, the pod's annotations update will cause a pod delete and re-create. This is not ideal.
Is there any way to just force the pod sync to make the config map changes visible to it?

**Comment by ThoTischner** [comment]
Hey @tangzhankun you need to update the annotation directly on the pod not the sts.

**Comment by rphillips** [comment]
Considering the original bug description, this issues seems to be fixed, since configmaps updates are more consistent and reliable. --sync-frequency (which can be configured to be more frequently to a palatable value, a user can use the workaround [described above](https://github.com/kubernetes/kubernetes/issues/30189#issuecomment-672889328) to refresh faster, and this has been documented.

As such, I believe we can close this. If people want to make the syncs more frequent or more reactive; perhaps we should consider a new issue with clear measurable requirements?

**Comment by ehashman** [comment]
/close

> If people want to make the syncs more frequent or more reactive; perhaps we should consider a new issue with clear measurable requirements?

Feel free to reopen if you want to pick this up.

**Comment by bo-er** [comment]
> Hey @tangzhankun you need to update the annotation directly on the pod not the sts. So you could add an annotation and immediately remove it, to trigger a pod sync that will sync the config map.

But what if I am updating a configmap from pod A, and if pod A restarts, my current operation is interrupted.  Is it possible to speed up without pod restarting?

**Comment by xzycn** [comment]
> Hey @tangzhankun you need to update the annotation directly on the pod not the sts. So you could add an annotation and immediately remove it, to trigger a pod sync that will sync the config map.

How about to deployment? Dose kubelet will sync configmap immediately when restart a deployment?

**Comment by zou2699** [comment]
> > Hey @tangzhankun you need to update the annotation directly on the pod not the sts. So you could add an annotation and immediately remove it, to trigger a pod sync that will sync the config map.
> 
> How about to deployment? Dose kubelet will sync configmap immediately when restart a deployment?
you need to update the annotation directly on the pod not the deploy or sts.

**Comment by xzycn** [comment]
@zou2699 
This is a weird way to do this，in most cases we  manage pods  through deployment or statefulset instead  of managing pods directly. If you update the annotation of a pod templdate, then deployment will rolling upgrade the existing pods, in this case,
dose kubelet will sync configmap immediately?

**Comment by zou2699** [comment]
> @zou2699 This is a weird way to do this，in most cases we manage pods through deployment or statefulset instead of managing pods directly. If you update the annotation of a pod templdate, then deployment will rolling upgrade the existing pods, in this case, dose kubelet will sync configmap immediately?

yes, but update the pod's annotation kubelet will not recreate pod, just reconcile it

**Comment by babaksit** [comment]
> > @zou2699 This is a weird way to do this，in most cases we manage pods through deployment or statefulset instead of managing pods directly. If you update the annotation of a pod templdate, then deployment will rolling upgrade the existing pods, in this case, dose kubelet will sync configmap immediately?
> 
> yes, but update the pod's annotation kubelet will not recreate pod, just reconcile it

In the case of Prometheus, I have a ServiceMonitor for a pod that I need to be synced immediately, should I update the pod's annotation which is being monitored or Prometheus pod ? and It could be any annotation right? Because when I update the pod's annotation it still takes around 1 minute to get updated
